### PR TITLE
Block function command line option

### PR DESCRIPTION
**Pull request by polgreen**
Blocks a function by replacing the function body with assume_false.
This is similar to --undefined-function-is-assume-false, but gives the flexibility to havoc undefined functions and block specific functions in one go without needing multiple calls to goto-instrument.

I still need to create a unit test and add it to this pull request, but any early feedback welcome

### CODE REVIEW

File: src/goto-instrument/remove_function.cpp
Comment (peterschrammel) [inline src/goto-instrument/remove_function.cpp]: Should be "Blocking"

File: src/goto-instrument/remove_function.cpp
Comment (peterschrammel) [inline src/goto-instrument/remove_function.cpp]: "removed" -> "blocked"

File: src/goto-instrument/remove_function.cpp
Comment (peterschrammel) [inline src/goto-instrument/remove_function.cpp]: run clang-format

File: src/goto-instrument/remove_function.cpp
Comment (peterschrammel) [inline src/goto-instrument/remove_function.cpp]: doxygen here

File: src/goto-instrument/goto_instrument_parse_options.cpp
Comment (thk123) [inline src/goto-instrument/goto_instrument_parse_options.cpp]: Consider adding `(may be repeated)` as above to indicate it can be used more than once.

**Review by thk123 — APPROVED** [review]
Seems good (pending the clang-format and doxygen changes).

**Comment by martin-cs** [comment]
Seems reasonable enough on first glance.  Ping me when the other comments have been applied and the "do not merge" has been removed.

File: src/goto-instrument/remove_function.cpp
Comment (chrisr-diffblue) [inline src/goto-instrument/remove_function.cpp]: It would be nice to have some Doxygen comments for this function, as @peterschrammel mentioned.

Comment (polgreen) [reply src/goto-instrument/remove_function.cpp]: I've now moved the doxygen from the .h file to the .cpp file now, although I understood that in general the rule was to put it in the header file?

File: src/goto-instrument/remove_function.cpp
Comment (smowton) [inline src/goto-instrument/remove_function.cpp]: typo modified

File: src/goto-instrument/remove_function.cpp
Comment (smowton) [inline src/goto-instrument/remove_function.cpp]: return values or side-effects

File: src/goto-instrument/remove_function.cpp
Comment (smowton) [inline src/goto-instrument/remove_function.cpp]: indent 2 I think

File: src/goto-instrument/remove_function.cpp
Comment (smowton) [inline src/goto-instrument/remove_function.cpp]: Braces around multi-line forall

File: src/goto-instrument/remove_function.cpp
Comment (smowton) [inline src/goto-instrument/remove_function.cpp]: Spaces around `=` / `==`

File: src/goto-instrument/remove_function.cpp
Comment (smowton) [inline src/goto-instrument/remove_function.cpp]: That's an odd comment. How about "...was explicitly marked unusable by command-line option `--block-function`"?

File: src/goto-instrument/remove_function.cpp
Comment (smowton) [inline src/goto-instrument/remove_function.cpp]: How about simply replacing the function body, rather than every callsite?

**Review by smowton — COMMENTED** [review]
Looks reasonable, some nitpicks on style. @peterschrammel @martin-cs do either of you know of standard terminology for this? "Block" made me think of "basic block" first and foremost. `--mark-function-unreachable` perhaps?

File: src/goto-instrument/remove_function.cpp
Comment (polgreen) [reply src/goto-instrument/remove_function.cpp]: So, originally I did that (have a look at https://github.com/diffblue/cbmc/pull/1566/commits/8708282e2b98cb26921dcddbbaf2d921bd8d1a44), but I got into difficulties when it started violating the asset at line 320 of local_bitvector_analysis.cpp: 
assert(succ<loc_infos.size())

I attempted to find a solution to this, and ended up deciding to mirror the code in undefined_function_abort_path (which is, in principle, doing a very similar thing)

Do you know how to solve the assertion violation? I assume that loc_infos needs updating somehow, but I wasn't able to work out how

**Comment by smowton** [comment]
Oh one more thing, should make the name of the cpp / h file consistent with the eventually chosen name for the pass

File: src/goto-instrument/remove_function.cpp
Comment (smowton) [reply src/goto-instrument/remove_function.cpp]: `loc_infos` is created when local-bitvector-analysis is run, *after* you've messed with the GOTO programs, so it won't need updating, but this does suggest the alteration somehow left the GOTO program in an inconsistent state. Couldn't say how without some deeper debugging though. I'd be happy to keep this as-is in the meantime, but it'd be a good cleanup to do if time permits.

**Comment by polgreen** [comment]
is "block-calls-to-function" more descriptive? I don't like "mark-function-unreachable" as to me that sounds too similar to "mark-unreachable-functions", I think "make-function-unreachable" might be better, but it's not clear how it is being made unreachable. 

The equivalent command for doing this for an undefined function is just "undefined-function-is-assume-false", so we could make this "make-function-assume-false"

**Comment by smowton** [comment]
make-function-assume-false is fine with me

**Comment by smowton** [comment]
Rather than merge develop into this, please rebase this branch on current develop

**Comment by polgreen** [comment]
Yeah, I'm on it. The last rebase was awful, some of the changes to the regression tests for java strings created conflicts in basically every single test.desc. In the end I gave up and checked out develop and cherry-picked my changes on it.

**Comment by polgreen** [comment]
I rebased this on the current branch when I made the last comment, what else is left for me to do?

**Comment by tautschnig** [comment]
I was just confused by the "I'm on it" I guess. Re-assigning to @peterschrammel to evaluate the "Do not merge" and possibly actually approve and merge.

**Comment by polgreen** [comment]
Sorry, I think I posted the comment slightly before I pushed the rebase and then just forgot to post an update

File: src/goto-instrument/goto_instrument_parse_options.cpp
Comment (smowton) [inline src/goto-instrument/goto_instrument_parse_options.cpp]: Typo?

**Review by smowton — APPROVED** [review]
Think the help text has got mangled, otherwise lgtm

**Comment by polgreen** [comment]
help text is fixed

**Comment by martin-cs** [comment]
I don't want to be That Guy but ... would this be best addressed as part of a comprehensive solution to : https://github.com/diffblue/cbmc/pull/1585 ?

**Comment by polgreen** [comment]
I think the suggested solution to https://github.com/diffblue/cbmc/pull/1585 was to create stub bodies for the functions and insert the changes into the stub bodies. 

Originally that is what I did for this PR and there is a discussion with @smowton about this in the comments above. I created stub bodies for the functions and inserted "assume(false)" into the stub bodies. This change was fine when I applied it to the version of CBMC I was using at the time. When I applied it to develop it caused violation of the assert at line 320 of local_bitvector_analysis.cpp:
assert(succ<loc_infos.size())

I couldn't fix this and decided it made sense to mirror the code in undefined_function_abort_path instead (which is doing a very similar thing). 

So, the options are:
- this pull request could be left consistent with undefined_function_abort_path
- this pull request could be changed to create a stub body and add instructions to it, but I would need some help debugging the assertion violation described above. It will then be inconsistent with undefined_function_abort_path. 
- all the possible ways of handling undefined functions could be changed to be consistent. I don't have time to do this before the end of January, so if this is the solution it's best it is assigned to someone else

**Comment by martin-cs** [comment]
Assorted thoughts...

1. The assertion failure looks like a corrupted goto-program.  The most obvious source is not calling update() after modifying it.  Without code it is harder to be clear.

2. Consistency, especially between similar tasks, if good but shouldn't keep us going down the wrong road (if we are).

3. Given the variety of behaviours that are potentially desirable for functions without bodies and the need for consistency between back-ends I think the "transform the program" approach is the right way forward.

I'm certainly happy to review, probably to help program even but it will have to be late January at the earliest.

**Pull request closed without merging**
